### PR TITLE
fix(core): Only assume empty square if reported square size is 1 *and* there are no txs

### PR DESCRIPTION
**Pull request by renaynay**
A bug was caught on blockspacerace network that caused a data root mismatch between bridge nodes and core nodes because celestia-node incorrectly assumes an empty square if there are no transactions which is not the case as squares can contain only tail-padding even if no transactions land in the block itself. 

We also will now: 
* rely on using `data.SquareSize` from the block data for extending shares instead of calculating square size from `len(shares)`, and
* panic on data root mismatch

**Future considerations:**

* we need a fuzzing suite against our block extension mechanism that can generate all kinds of blocks (our integration testing doesn't test for edge cases like this.
* we need to audit the rest of the codebase to ensure we perform the same calculations over square size or assume empty square where we shouldn't be.

### CODE REVIEW

**Review by distractedm1nd — DISMISSED** [review]

**Review by vgonkivs — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

File: core/eds.go [hunk 1]
```diff
@@ -20,15 +19,15 @@ import (
...
 		return nil, nil
 	}
 	shares, err := appshares.Split(data, true)
 	if err != nil {
 		return nil, err
 	}
-	size := utils.SquareSize(len(shares))
-	return da.ExtendShares(size, appshares.ToBytes(shares))
+
+	return da.ExtendShares(data.SquareSize, appshares.ToBytes(shares))
```
Comment (evan-forbes) [inline core/eds.go:30]: I think I was wrong to say that this is necessary, but this is a good change anyway imo

**Review by Wondertan — APPROVED** [review]

File: header/header.go [hunk 2]
```diff
@@ -141,8 +141,8 @@ func (eh *ExtendedHeader) Validate() error {
 
 	// ensure data root from raw header matches computed root
 	if !bytes.Equal(eh.DAH.Hash(), eh.DataHash) {
-		return fmt.Errorf("mismatch between data hash commitment from core header and computed data root: "+
-			"data hash: %X, computed root: %X", eh.DataHash, eh.DAH.Hash())
+		panic(fmt.Sprintf("mismatch between data hash commitment from core header and computed data root "+
+			"at height %d: data hash: %X, computed root: %X", eh.Height(), eh.DataHash, eh.DAH.Hash()))
```
Comment (evan-forbes) [inline header/header.go:145]: [very optional]

stopping the node is the correct thing to do here imo, but perhaps not showing the user a stack trace would be good

**Review by evan-forbes — APPROVED** [review]

**Review by walldiss — APPROVED** [review]

File: core/eds.go (hunk 1 above)
Comment (cmwaters) [reply core/eds.go:30]: Yeah I think these should always be the same number. But better probably to use `data.SquareSize` from a defensive perspective
